File: cpp/src/solver.cpp
#include "solver.hpp"
#include "cell.hpp"
#include "tree.hpp"
#include "utils.hpp"
#include <algorithm>
#include <boost/qvm.hpp>
#include <boost/qvm/math.hpp>
#include <fmt/base.h>
#include <ranges>
#include <thread>

namespace qvm = boost::qvm;

GenericSolver::GenericSolver(
  const double dt,
  const double epsilon,
  const std::vector<MassSample>& samples,
  const std::function<double(const Vec3&)>& phi,
  const std::function<Vec3(const Vec3&)>& grad_phi,
  const double G)
  : dt(dt)
  , epsilon(epsilon)
  , samples(samples)
  , phi(phi)
  , grad_phi(grad_phi)
  , G(G)
{}

double
GenericSolver::potential(const Vec3& diff) const
{
  return -G / qvm::mag(diff) * phi(diff / epsilon);
}

Vec3
GenericSolver::field_intensity(const Vec3& diff) const
{
  return G / qvm::mag_sqr(diff) * grad_phi(diff / epsilon);
}

Vec3
GenericSolver::average_position() const
{
  Vec3 avg{ 0.0f };
  for (const MassSample& sample : samples)
    avg += sample.position;
  return avg / static_cast<double>(samples.size());
}

Vec3
GenericSolver::std_pos() const
{
  using qvm::X, qvm::Y, qvm::Z;

  const Vec3 avg = average_position();
  Vec3 std{ 0.0f };
  for (const MassSample& sample : samples) {
    const Vec3 diff = sample.position - avg;
    std += Vec3{ X(diff) * X(diff), Y(diff) * Y(diff), Z(diff) * Z(diff) };
  }
  std /= static_cast<double>(samples.size());
  X(std) = std::sqrt(X(std));
  Y(std) = std::sqrt(Y(std));
  Z(std) = std::sqrt(Z(std));
  return std;
}

double
GenericSolver::pos_divergence(const GenericSolver& other) const
{
  const auto& range = std::ranges::views::iota;

  double divergence = 0.0f;
  for (size_t i : range(0ul, samples.size())) {
    const Vec3 diff = samples[i].position - other.samples[i].position;
    divergence += qvm::mag_sqr(diff);
  }
  divergence /= static_cast<double>(samples.size());
  return divergence;
}

double
GenericSolver::total_energy() const
{
  // Kinetic energy
  double ke = 0.0f;
  for (const MassSample& sample : samples) {
    ke += 0.5f * sample.mass * qvm::mag_sqr(sample.speed(dt));
  }
  // Potential energy
  double pe = 0.0f;
  for (const MassSample& s1 : samples)
    for (const MassSample& s2 : samples)
      if (&s1 != &s2)
        pe += potential(s1.position - s2.position) * s1.mass * s2.mass;
  return ke + pe;
}

FMMSolver::FMMSolver(
  const double size,
  const double dt,
  const std::vector<MassSample>& samples,
  const index_t depth,
  const std::function<double(const Vec3&)>& phi,
  const std::function<Vec3(const Vec3&)>& grad_phi,
  const std::optional<std::function<Mat3x3(const Vec3&)>>& hess_phi,
  const double G)
  : GenericSolver(
      dt,
      4.0f * size / std::sqrt(static_cast<double>(samples.size())),
      samples,
      phi,
      grad_phi,
      G)
  , tree(FMMTree(depth, size))
  , size(size)
  , hess_phi(hess_phi)
{}

Mat3x3
FMMSolver::field_jacobian(const Vec3& diff) const
{
  if (hess_phi.has_value())
    return G / std::pow(qvm::mag(diff), 3.0f) *
           hess_phi.value()(diff / epsilon);
  else
    return Mat3x3{ 0.0f };
}

Vec3
FMMSolver::compute_close(const FMMCell& cell, const MassSample& sample) const
{
  Vec3 total_field = { 0.0f };
  for (const auto& neighbor_cell : cell.direct_neighbors) {
    if (neighbor_cell.get().samples.empty())
      continue;
    for (const auto& neighbor_sample : neighbor_cell.get().samples) {
      const Vec3 diff = sample.position - neighbor_sample.get().position;
      if (qvm::mag(diff) == 0.0f)
        continue;
      total_field += neighbor_sample.get().mass * field_intensity(diff);
    }
  }
  return total_field;
}

Vec3
FMMSolver::compute_far(const FMMCell& cell, const MassSample& sample)
{
  return cell.field_tensor.field +
         cell.field_tensor.jacobian * (sample.position - cell.barycenter);
}

void
FMMSolver::update(const size_t threads)
{
  const auto& range = std::ranges::views::iota;

  tree.update(
    samples,
    [this](const Vec3& diff) { return field_intensity(diff); },
    [this](const Vec3& diff) { return field_jacobian(diff); });

  std::vector<Vec3> new_poss(samples.size());

  index_t index = 0;
  const index_t l = tree.depth() - 1, width = 1 << l;
  for (const index_t i : range(0, width)) {
    for (const index_t j : range(0, width)) {
      for (const index_t k : range(0, width)) {
        const FMMCell& cell = tree[l][i][j][k];
        if (cell.samples.empty())
          continue;
        for (const auto& sample : cell.samples) {
          Vec3 acc = { 0.0f };
          acc += compute_close(cell, sample);
          acc += compute_far(cell, sample);
          new_poss[index] =
            2 * sample.get().position - sample.get().prev_pos + acc * dt * dt;
          index++;
        }
      }
    }
  }

  index = 0;
  for (const index_t i : range(0, width)) {
    for (const index_t j : range(0, width)) {
      for (const index_t k : range(0, width)) {
        const FMMCell& cell = tree[l][i][j][k];
        if (cell.samples.empty())
          continue;
        for (auto& sample : cell.samples) {
          sample.get().prev_pos = sample.get().position;
          sample.get().position = new_poss[index];
          index++;
        }
      }
    }
  }
}

void
NaiveSolver::update(const size_t threads)
{
  std::vector<Vec3> new_poss;
  new_poss.resize(samples.size());
  std::ranges::fill(new_poss, Vec3{ 0.0f, 0.0f, 0.0f });

  for (index_t index = 0; index < samples.size(); ++index) {
    Vec3 acc = { 0.0f };
    const MassSample& sample1 = samples[index];
    for (const MassSample& sample2 : samples) {
      const Vec3& diff = sample1.position - sample2.position;
      if (qvm::mag_sqr(diff) != 0.0f)
        acc += sample2.mass * field_intensity(diff);
    }
    new_poss[index] = 2 * sample1.position - sample1.prev_pos + acc * dt * dt;
  }

  for (index_t index = 0; index < samples.size(); index++) {
    MassSample& sample = samples[index];
    sample.prev_pos = sample.position;
    sample.position = new_poss[index];
  }
}

File: src/solver.py
from __future__ import annotations

from typing import Callable, Generator, List

import numpy as np

from .cell import FMMCell
from .sample import MassSample
from .tree import FMMTree
from .utils import Mat3x3, Vec3


class GenericSolver:
    """
    Generic solver template for particles systems physics.

    Fields:

        phi: function of r/epsilon representing the potential
            of one "particle" per unit mass (Callable[[Vec3], float])

        grad_phi: function of r/epsilon representing the gradient
            potential of one "particle" per unit mass (Callable[[Vec3], Vec3])

        dt: timestep (float)

        samples: list of mass samples (MassSample)

        epsilon: particle smoothing size (float)

        G: gravitational constant (float)
    """

    def __init__(
        self,
        dt: float,
        epsilon: float,
        samples: List[MassSample],
        phi: Callable[[Vec3], float],
        grad_phi: Callable[[Vec3], Vec3],
        G: float = 1e-1,
    ):
        self.phi = phi
        self.grad_phi = grad_phi
        self.dt = dt
        self.samples = samples
        self.epsilon = epsilon
        self.G = G

    def potential(self, diff: Vec3) -> float:
        """
        Helper function to compute the potential between two points, given their position difference.

        Args:
            diff: position difference between the two points (Vec3)

        Returns:
            potential (float)
        """
        return float(-self.G / np.linalg.norm(diff) * self.phi(diff / self.epsilon))

    def field_intensity(self, diff: Vec3) -> Vec3:
        """
        Helper function to compute the field intensity from one point to another, given their position difference.

        Args:
            diff: position difference between the two points (Vec3)

        Returns:
            field intensity (Vec3)
        """
        return self.G / (np.linalg.norm(diff) ** 2) * self.grad_phi(diff / self.epsilon)

    def update(self):
        """
        Updates the system.
        """
        pass

    # Utilities

    def average_pos(self) -> Vec3:
        """
        Average position of the samples.
        """
        avg: Vec3 = np.zeros(3)
        for s in self.samples:
            avg += s.pos
        return avg / len(self.samples)

    def std_pos(self) -> Vec3:
        """
        Standard deviation of the position of the samples.
        """
        avg_sq: Vec3 = np.zeros(3)
        avg = self.average_pos()
        for s in self.samples:
            avg_sq += (s.pos - avg) ** 2
        return np.sqrt(avg_sq / len(self.samples))

    def pos_divergence(self, lhs: GenericSolver) -> float:
        """
        Position divergence between two solvers.
        """
        div = 0
        for sample1, sample2 in zip(self.samples, lhs.samples):
            div += ((sample1.pos - sample2.pos) ** 2).sum()
        return div / len(self.samples)

    def total_energy(self) -> float:
        """
        Total energy of the system.
        """
        # Kinetic energy
        ke = sum(
            s.mass * np.linalg.norm(s.speed(self.dt)) ** 2 / 2 for s in self.samples
        )
        # Potential energy
        pe = 0
        for s1 in self.samples:
            for s2 in self.samples:
                if s1 is not s2:
                    pe += self.potential(s1.pos - s2.pos) * s1.mass * s2.mass
        return float(ke + pe)


class FMMSolver(GenericSolver):
    """
    Solver for particle systems using the Fast Multipole Multiplication method.
    If the depth of the space tree is chosen to be O(log8(len(samples)),
    the time complexity of the algorithm is expected to be O(len(samples)).

    Fields:

        size: size of the simulation cube (float)

        phi: function of r/epsilon representing the potential
            of one "particle" per unit mass (Callable[[Vec3], float])

        grad_phi: function of r/epsilon representing the gradient
            potential of one "particle" per unit mass (Callable[[Vec3], Vec3])

        hess_phi: function of r/epsilon representing the hessian
            potential of one "particle" per unit mass. Set to None to
            use an order 0 expansion of the inter-cell field. None by default.
            (Callable[[Vec3], Mat3x3] | None)

        dt: timestep (float)

        tree: octree of the volume cells (FMMTree)

        samples: list of mass samples (MassSample)

        epsilon: particle smoothing size (float)

        G: gravitational constant (float, defaults to 0.1)
    """

    def __init__(
        self,
        size: float,
        dt: float,
        samples: List[MassSample],
        depth: int,
        phi: Callable[[Vec3], float],
        grad_phi: Callable[[Vec3], Vec3],
        hess_phi: Callable[[Vec3], Mat3x3] | None = None,
        G: float = 1e-1,
    ):
        self.size = size
        self.phi = phi
        self.grad_phi = grad_phi
        self.hess_phi = hess_phi
        self.dt = dt
        self.tree = FMMTree(depth, size)
        self.samples = samples
        self.epsilon = 4 * size / np.sqrt(len(samples))
        self.G = G

    def field_jacobian(self, diff: Vec3) -> Mat3x3:
        """
        Helper function to compute the field jacobian from one point to another, given their position difference.

        Args:
            diff: position difference between the two points (Vec3)

        Returns:
            field jacobian (Mat3x3)
        """
        return (
            np.zeros((3, 3))
            if self.hess_phi is None
            else (
                self.G
                / (np.linalg.norm(diff) ** 3)
                * self.hess_phi(diff / self.epsilon)
            )
        )

    def update(self):
        """
        Updates the system with the FMM algorithm, with time step dt.

        The algorithm starts by updating the samples' distribution in the tree
        structure, then updates the multipoles, barycenters andfield tensors on each cell,
        and finally computes the field intensity at each point to update the positions
        using the Verlet integration scheme.

        Time complexity:
            O(8^self.tree.depth() + len(self.samples)), which becomes O(len(self.samples))
            if the depth of the tree is chosen to be O(log8(len(samples)).
        """
        self.tree.update(self.samples, self.field_intensity, self.field_jacobian)
        new_poss = np.zeros((len(self.samples), 3))
        index = 0
        depth = self.tree.depth()
        for i in range(2 ** (depth - 1)):
            for j in range(2 ** (depth - 1)):
                for k in range(2 ** (depth - 1)):
                    cell = self.tree[-1][i][j][k]
                    assert cell.samples is not None
                    if len(cell.samples) == 0:
                        continue
                    for sample in cell.samples:
                        grad_potential = np.zeros(3)
                        grad_potential += self.compute_close(cell, sample)
                        grad_potential += self.compute_far(cell, sample)
                        acc = grad_potential
                        new_poss[index] = (
                            2 * sample.pos - sample.prev_pos + acc * self.dt**2
                        )
                        index += 1

        index = 0
        for i in range(2 ** (depth - 1)):
            for j in range(2 ** (depth - 1)):
                for k in range(2 ** (depth - 1)):
                    cell = self.tree[-1][i][j][k]
                    assert cell.samples is not None
                    if len(cell.samples) == 0:
                        continue
                    for sample in cell.samples:
                        sample.prev_pos = sample.pos
                        sample.pos = new_poss[index]
                        index += 1

    def compute_close(self, cell: FMMCell, sample: MassSample) -> Vec3:
        total_field = np.zeros(3)
        for neighbor_cell in cell.direct_neighbors:
            assert neighbor_cell.samples is not None
            if len(neighbor_cell.samples) == 0:
                continue
            for close_sample in neighbor_cell.samples:
                diff = sample.pos - close_sample.pos
                if np.linalg.norm(diff) == 0:
                    continue
                field_intensity = self.field_intensity(diff)
                total_field += close_sample.mass * field_intensity
        return total_field

    def compute_far(self, cell: FMMCell, sample: MassSample) -> Vec3:
        return cell.field_tensor.field + cell.field_tensor.jacobian @ (
            sample.pos - cell.barycenter
        )

    def iter_cells(self, floor: int) -> Generator[FMMCell]:
        if floor >= self.tree.depth():
            raise ValueError(
                f"Asked floor {floor} in a tree of depth {self.tree.depth()}."
                "Floor must be less than the depth of the tree."
            )
        floor %= self.tree.depth()
        for i in range(2**floor):
            for j in range(2**floor):
                for k in range(2**floor):
                    yield self.tree[floor][i][j][k]


class NaiveSolver(GenericSolver):
    """
    Simple quadratic complexity solver for particle systems.
    Used for speed comparison with better algorithms.

    Fields:

        phi: function of r/epsilon representing the potential
            of one "particle" per unit mass (Callable[[Vec3], float])

        grad_phi: function of r/epsilon representing the gradient
            potential of one "particle" per unit mass (Callable[[Vec3], Vec3])

        dt: timestep (float)

        samples: list of mass samples (MassSample)

        epsilon: particle smoothing size (float)

        G: gravitational constant (float)
    """

    def update(self):
        """
        Updates the system with the basic O(len(self.samples)^2) algorithm.
        Used for speed comparison with the enhanced algorithm.
        """
        new_poss = np.zeros((len(self.samples), 3))
        for index, sample1 in enumerate(self.samples):
            acc = np.zeros(3)
            for sample2 in self.samples:
                diff = sample1.pos - sample2.pos
                if np.linalg.norm(diff) == 0:
                    continue
                field_intensity = self.field_intensity(diff)
                acc += sample2.mass * field_intensity
            new_poss[index] = 2 * sample1.pos - sample1.prev_pos + acc * self.dt**2

        for index, sample in enumerate(self.samples):
            sample.prev_pos = sample.pos
            sample.pos = new_poss[index]
